### PR TITLE
feat: add lora and qlora

### PR DESCRIPTION
**Pull request by jemmyshin**
add lora and qlora

### CODE REVIEW

File: open_gpt/cli/commands/serve.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ class ServeCommand(Command):
         option(
             'precision', None, 'The precision of the model.', flag=False, default='fp16'
         ),
+        option('peft_name', None, 'The name of the LORA model.'),
```
Comment (numb3r3) [inline open_gpt/cli/commands/serve.py:30]:
```suggestion
        option('lora_name_or_path', None, 'The name or path of the LORA checkpoint.'),
```

File: open_gpt/cli/commands/serve.py [hunk 2]
```diff
@@ -56,6 +57,7 @@ def handle(self) -> int:
             cors=self.option('enable_cors'),
             uses_with={
                 'precision': self.option('precision'),
+                'peft_model_id_or_path': self.option('peft_name'),
```
Comment (numb3r3) [inline open_gpt/cli/commands/serve.py:60]:
```suggestion
                'lora_name_or_path': self.option('lora_name_or_path'),
```

File: open_gpt/factory.py [hunk 3]
```diff
@@ -7,13 +7,15 @@
 def create_model(
     model_name: str,
     precision: str = 'fp32',
+    peft_model_id_or_path: Optional[str] = None,
```
Comment (numb3r3) [inline open_gpt/factory.py:10]: same

File: open_gpt/models/flamingo/loading.py [hunk 4]
```diff
@@ -11,6 +11,7 @@ def load_model_and_transforms(
     model_name_or_path: str,
     vision_model_name_or_path: str,
     lang_model_name_or_path: str,
+    peft_model_id_or_path: Optional[str] = None,
```
Comment (numb3r3) [inline open_gpt/models/flamingo/loading.py:14]: I don't think we have lora accessible for flamingo so far. Please revert it.

File: open_gpt/models/llama/loading.py [hunk 5]
```diff
@@ -73,4 +74,25 @@ def load_model_and_tokenizer(
...
+    """
+    Use LORA for inference, `peft_model_id_or_path` can be either an id on huggingface 
+    or the name of a directory. If it's a directory, there should be two files in it:
+    checkpoint file which contains the weight of LORA, named `adapter_model.bin`,
+    config file which will be used to initialize LORA, named `adapter_config.json`.
+    ** The file names cannot be changed **
+    
+    For details about PeftModel.from_pretrained, see:
+    `peft.utils.peft_model.py::load_adapter()` for loading ckpt,
+    `peft.utils.config.py::from_pretrained()` for loading config.
```
Comment (numb3r3) [inline open_gpt/models/llama/loading.py:86]: Rephrase the docstr for this API with more details. It is weird just to mention lora here.

File: open_gpt/models/llama/modeling.py [hunk 6]
```diff
@@ -12,13 +12,17 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def load_model_and_transforms(
-        self, model_name_or_path: str, tokenizer_name_or_path: Optional[str] = None
+        self,
+        model_name_or_path: str,
+        peft_model_id_or_path: Optional[str] = None,
```
Comment (numb3r3) [inline open_gpt/models/llama/modeling.py:17]: change para name

File: open_gpt/models/loading.py [hunk 7]
```diff
@@ -75,4 +76,25 @@ def load_model_and_tokenizer(
...
+    """
+    Use LORA for inference, `peft_model_id_or_path` can be either an id on huggingface 
+    or the name of a directory. If it's a directory, there should be two files in it:
+    checkpoint file which contains the weight of LORA, named `adapter_model.bin`,
+    config file which will be used to initialize LORA, named `adapter_config.json`.
+    ** The file names cannot be changed **
+
+    For details about PeftModel.from_pretrained, see:
+    `peft.utils.peft_model.py::load_adapter()` for loading ckpt,
+    `peft.utils.config.py::from_pretrained()` for loading config.
```
Comment (numb3r3) [inline open_gpt/models/loading.py:88]: same, please turn on copilot for this task!!!!!

File: open_gpt/models/loading.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 
 def load_model_and_tokenizer(
```
Comment (numb3r3) [inline open_gpt/models/loading.py:8]: I would suggest to have a mixin function `load_lora(self)` which can be used by different model wrapper without copy codes.

File: pyproject.toml [hunk 9]
```diff
@@ -55,11 +55,12 @@ cleo = "^2.0.0"
 click = "^8.1.3"
 numpy = "^1.21.2"
 einops = "^0.6.0"
-transformers = "^4.29.0"
+transformers = { git = "https://github.com/huggingface/transformers.git", branch = "main" }
```
Comment (numb3r3) [inline pyproject.toml:58]: reject, I knew it only works on the latest transformer. But I would not mention it in our dependency.

**Review by numb3r3 — CHANGES_REQUESTED** [review]
use a mixin function `load_lora(self, ...)` instead.

File: pyproject.toml (hunk 9 above)
Comment (numb3r3) [reply pyproject.toml:58]: just wait for a while the main branch is released. Then, we just bump the tag version.

**Comment by numb3r3** [comment]
And it does not necessary to make lora works for all models in this PR.

File: examples/llama_example.py [hunk 10]
```diff
@@ -17,9 +17,32 @@
...
+    peft_model_id_or_path='./my-lora',
+    device_map='balanced',
+)
+# test qlora using local ckpt file
 model = open_gpt.create_model(
-    'decapoda-research/llama-7b-hf', precision='fp16', device_map='balanced'
+    'decapoda-research/llama-7b-hf',
+    precision='bit4',
+    peft_model_id_or_path='./my-qlora',
+    device_map='balanced',
```
Comment (numb3r3) [inline examples/llama_example.py:43]: comment the unused codes. And don't add the debug code in our repo

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: open_gpt/models/vicuna/loading.py [hunk 11]
```diff
@@ -55,7 +57,10 @@ def load_model_and_tokenizer(
         **kwargs,
     )
 
-    logger.info(f"Loading model weights delta from {model_name_or_path}")
+    logger.info(
+        f"Loading model weights delta from {model_name_or_path}, "
+        f"loading LORA weights from {peft_model_id_or_path}"
+    )
```
Comment (numb3r3) [inline open_gpt/models/vicuna/loading.py:63]: revert

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: open_gpt/models/loading.py (hunk 8 above)
Comment (jemmyshin) [reply open_gpt/models/loading.py:8]: how about using a decorator `@load_lora()` to do this?

File: open_gpt/cli/commands/serve.py [hunk 12]
```diff
@@ -56,6 +59,7 @@ def handle(self) -> int:
             cors=self.option('enable_cors'),
             uses_with={
                 'precision': self.option('precision'),
+                'adapter_name_or_path': self.option('adapter_name_or_path'),
```
Comment (numb3r3) [inline open_gpt/cli/commands/serve.py:62]: The executor does not recognize this param yet. Please also update it in this PR.

File: open_gpt/factory.py [hunk 13]
```diff
@@ -27,6 +29,7 @@ def create_model(
 
         return LlamaModel(
             model_name,
+            adapter_name_or_path,
```
Comment (numb3r3) [inline open_gpt/factory.py:32]: It seems that only Llama supports lora checkpoint, please assert when the lora param is given on other model types.

File: open_gpt/models/modeling.py [hunk 14]
```diff
@@ -57,9 +62,23 @@ def load_model_and_transforms(
...
     def post_init(self, **kwargs):
         pass
+
+    def load_adapter(self, adapter_name_or_path: str):
+        from peft import PeftModel
+
+        self.model = PeftModel.from_pretrained(
+            self.model,
+            adapter_name_or_path,
+            device_map={'': self._device or 0}
```
Comment (numb3r3) [inline open_gpt/models/modeling.py:81]: `device_map` is also supported in peft?

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: open_gpt/models/llama/loading.py [hunk 15]
```diff
@@ -12,10 +12,23 @@ def load_model_and_tokenizer(
...
+    :param model_name_or_path: The model id or path to load the model from.
+    :param tokenizer_name_or_path: The tokenizer id or path to load the tokenizer from.
+    :param adapter_name_or_path: The adapter model name or path to load the adapter model from.
+    (`str` or `os.PathLike`)
+    The name of the adapter configuration to use. Can be either:
+    - A string, the `model id` of a adapter configuration hosted inside a model repo on the Hugging Face
+      Hub.
+    - A path to a directory containing a adapter configuration file saved using the `save_pretrained`
+      method (`./my_adapter_config_directory/`).
+
```
Comment (numb3r3) [inline open_gpt/models/llama/loading.py:28]: remove them because I think they are not well formated, and cannot be well rendered by docstr tool.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: examples/llama_example.py [hunk 16]
```diff
@@ -20,6 +20,29 @@
...
+#     precision='bit8',
+#     adapter_name_or_path='./my-lora',
+#     device_map='balanced',
+# )
+# llama + QLoRA using local ckpt file
+# model = open_gpt.create_model(
+#     'decapoda-research/llama-7b-hf',
+#     precision='bit4',
+#     adapter_name_or_path='./my-qlora',
+#     device_map='balanced',
```
Comment (numb3r3) [inline examples/llama_example.py:43]: I would suggest you use a separate new example `llama_lora_example.py` to make the example more clear and easy to understand.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: open_gpt/factory.py (hunk 13 above)
Comment (jemmyshin) [reply open_gpt/factory.py:32]: yes, but it's in `kwargs`, will be used in `load_adapter`: 
image

File: open_gpt/models/modeling.py (hunk 14 above)
Comment (jemmyshin) [reply open_gpt/models/modeling.py:81]: yes, but it's in `kwargs`, will be used in `load_adapter`: 
image

**Review by numb3r3 — APPROVED** [review]
LGTM

**Pull request merged**